codec/src/main/java/io/netty/handler/codec/CodecOutList.java

```diff
@@ -0,0 +1,182 @@
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.handler.codec;
+
+import io.netty.util.Recycler;
+
+import java.util.AbstractList;
+import java.util.RandomAccess;
+
+import static io.netty.util.internal.ObjectUtil.checkNotNull;
+
+/**
+ * Special {@link AbstractList} implementation which is used within our codec base classes.
+ */
+final class CodecOutList extends AbstractList<Object> implements RandomAccess {
+
+    private static final Recycler<CodecOutList> RECYCLER = new Recycler<CodecOutList>() {
+        @Override
+        protected CodecOutList newObject(Handle<CodecOutList> handle) {
+            return new CodecOutList(handle);
+        }
+    };
+
+    static CodecOutList newInstance() {
+        return RECYCLER.get();
+    }
+
+    private final Recycler.Handle<CodecOutList> handle;
+    private int size;
+    // Size of 16 should be good enough for 99 % of all users.
+    private Object[] array = new Object[16];
+    private boolean insertSinceRecycled;
+
+    private CodecOutList(Recycler.Handle<CodecOutList> handle) {
+        this.handle = handle;
+    }
+
+    @Override
+    public Object get(int index) {
+        checkIndex(index);
+        return array[index];
+    }
+
+    @Override
+    public int size() {
+        return size;
+    }
+
+    @Override
+    public boolean add(Object element) {
+        checkNotNull(element, "element");
+        try {
+            insert(size, element);
+        } catch (IndexOutOfBoundsException ignore) {
+            // This should happen very infrequently so we just catch the exception and try again.
+            expandArray();
+            insert(size, element);
+        }
+        size ++;
+        return true;
+    }
+
+    @Override
+    public Object set(int index, Object element) {
+        checkIndex(index);
+        checkNotNull(element, "element");
+
+        Object old = getAndAssert(index);
+
+        insert(index, element);
+        return old;
+    }
+
+    @Override
+    public void add(int index, Object element) {
+        checkIndex(index);
+        checkNotNull(element, "element");
+
+        if (size == array.length) {
+            expandArray();
+        }
+
+        if (index != size - 1) {
+            System.arraycopy(array, index, array, index + 1, size - index);
+        }
+
+        insert(index, element);
+        size ++;
+    }
+
+    @Override
+    public Object remove(int index) {
+        checkIndex(index);
```

Comment:
accessing the underlying array should do the bounds check ... do we need to duplicate the check?